### ISSUE TITLE
Flattening certain cubic béziers does not work

### ISSUE THREAD

**Issue by dbrgn** [nical/lyon#19]
Flattening this path results in a straight line:

    m 11.71726,9.07143 c -9.827381,4.15774 6.425594,10.20536 6.425594,10.20536

Result:

![img1](https://tmp.dbrgn.ch/screenshots/screenshot-20170406225638-kqzrsg55.png)

```
Segment: Relative MoveTo { x: 11.71726, y: 9.07143 }
Segment: Relative CurveTo { x1: -9.827381, y1: 4.15774, x2: 6.425594, y2: 10.20536, x: 6.425594, y: 10.20536 }
From (11.71726,9.07143) to (18.142855,19.27679) via (1.889879,13.22917) and (18.142855,19.27679)
 o 18.142855,19.27679
```

But when tweaking it just a little bit (e.g. changing the first value from `11.7...` to `11.8...`) it works:

    m 11.81726,9.07143 c -9.827381,4.15774 6.425594,10.20536 6.425594,10.20536

![img2](https://tmp.dbrgn.ch/screenshots/screenshot-20170406225804-uc3lghgs.png)

```
Segment: Relative MoveTo { x: 11.81726, y: 9.07143 }
Segment: Relative CurveTo { x1: -9.827381, y1: 4.15774, x2: 6.425594, y2: 10.20536, x: 6.425594, y: 10.20536 }
From (11.81726,9.07143) to (18.242853,19.27679) via (1.989879,13.22917) and (18.242853,19.27679)
 o 9.218064,10.766278
 o 8.6632,12.285181
 o 9.291998,13.824333
 o 11.29783,15.699478
 o 15.09445,17.905602
 o 18.242853,19.27679
```

The second control point and the end are equal, not sure if that causes an edge case.

**Comment by dbrgn** [comment]
(Sorry, submitted too early. Updated the issue text now with more details.)

**Comment by nical** [comment]
Thanks for filing this, I'll try to find the time to look into this this week.

**Comment by nical** [comment]
@dbrgn what was the value of the tolerance threshold you used to reproduce this bug?

**Comment by dbrgn** [comment]
@nical good point. I believe it was 0.15.

**Comment by nical** [comment]
I have tried to reproduce this by walking over the space of tolerance thresholds from 0.0000001 to 1.5 by increment of 0.0000001. And I haven't managed to get the flattening code to generate a straight line (see test below).

```rust
#[test]
fn test_issue_19() {
    let mut tolerance = 0.0000001;
    while tolerance < 1.5 {
        let c1 = CubicBezierSegment {
            from: Point::new(11.71726, 9.07143),
            ctrl1: Point::new(-9.827381, 4.15774),
            ctrl2: Point::new(6.425594, 10.20536),
            to: Point::new(6.425594, 10.20536),
        };
        let iter_points: Vec<Point> = c1.flattening_iter(tolerance).collect();
        let mut builder_points = Vec::new();
        c1.flattened_for_each(tolerance, &mut|p|{ builder_points.push(p); });

        assert_approx_eq(&iter_points[..], &builder_points[..]);

        if iter_points.len() <= 1 {
            print!(" tolerance {}", tolerance);
            print!(" points {:?}", iter_points);
            panic!();
        }

        tolerance += 0.0000001;
    }
}
```

I also tried to see it with the stroke tessellator using the example renderer and some hand picked threshold values (unfortunately can't try a gazillion of them manually). I did find a bug in the tessellator that is independent of the tolerance value, and looks unrelated to your issue.

So there must be something else rounding/nudging/affecting the coordinates or threshold before flattening or whatnot. @dbrgn do you have some code online that I can look at to see what else could be affecting your test case?
Which version of the crate are you using exactly?

**Comment by dbrgn** [comment]
@nical I'll try to reproduce it, will let you know once I have a working minimal example :)

**Comment by dbrgn** [comment]
Cargo.toml:

```toml
[package]
name = "straightline"
version = "0.1.0"
authors = ["Danilo Bargen <mail@dbrgn.ch>"]

[dependencies]
lyon_bezier = "^0.4"
```

main.rs:

```rust
extern crate lyon_bezier;

use lyon_bezier::{CubicBezierSegment, Vec2};

const FLATTENING_TOLERANCE: f32 = 0.15;

fn main() {
    println!("Hello, world!");

    let segment1 = CubicBezierSegment {
        from: Vec2::new(11.71726, 9.07143),
        ctrl1: Vec2::new(1.889879,13.22917),
        ctrl2: Vec2::new(18.142855,19.27679),
        to: Vec2::new(18.142855,19.27679),
    };
    println!("\nSegment: {:?}", segment1);
    for point in segment1.flattening_iter(FLATTENING_TOLERANCE) {
        println!("Point: {:?}", point);
    }

    let segment2 = CubicBezierSegment {
        from: Vec2::new(11.71726, 9.07143),
        ctrl1: Vec2::new(1.889879,13.22917),
        ctrl2: Vec2::new(18.142855,19.27679),
        to: Vec2::new(18.142855,19.27678),
    };
    println!("\nSegment: {:?}", segment2);
    for point in segment2.flattening_iter(FLATTENING_TOLERANCE) {
        println!("Point: {:?}", point);
    }

}
```

(Difference in segment1 and segment2: The last digit of `to`)

Output:

```
Hello, world!

Segment: CubicBezierSegment { from: (11.71726,9.07143), ctrl1: (1.889879,13.22917), ctrl2: (18.142855,19.27679), to: (18.142855,19.27679) }
Point: (18.142855,19.27679)

Segment: CubicBezierSegment { from: (11.71726,9.07143), ctrl1: (1.889879,13.22917), ctrl2: (18.142855,19.27679), to: (18.142855,19.27678) }
Point: (9.118064,10.766278)
Point: (8.563201,12.285181)
Point: (9.191999,13.824334)
Point: (11.197831,15.699478)
Point: (14.99445,17.905596)
Point: (18.142855,19.27678)
```

**Comment by nical** [comment]
Thanks for the test case, it looks like some precision issue in find_cubic_bezier_inflection_approximation_range, I need to delve back into the maths to understand why that is but it is indeed related to the control points being very close to one another.

**Comment by nical** [comment]
Equivalent bug filed against Gecko (this code being a straightforward port of Gecko's path flattening code): https://bugzilla.mozilla.org/show_bug.cgi?id=1360143

**Comment by nical** [comment]
Fixed in #20.
